### PR TITLE
Crashfix newtagfilter

### PR DESCRIPTION
**Pull request by JB184351**

### CODE REVIEW

File: TalkAloud/TalkAloud/Cells/TagCell/TagCell.swift [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import UIKit
+
+protocol TagFilterDelegate: class {
+    func didUpdateTagToFilter(by tags: [TagModel])
+}
+
+class TagCell: UITableViewCell {
+
+    @IBOutlet var collectionView: UICollectionView!
+    private var selectedTags = [String]()
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/Cells/TagCell/TagCell.swift]: It seems that you don't need this property anymore

File: TalkAloud/TalkAloud/Cells/TagCell/TagCell.swift [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+//  Copyright © 2020 Justin Bengtson. All rights reserved.
+//
+
+import UIKit
+
+protocol TagFilterDelegate: class {
+    func didUpdateTagToFilter(by tags: [TagModel])
+}
+
+class TagCell: UITableViewCell {
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/Cells/TagCell/TagCell.swift:15]: Add all the // MARK:

File: TalkAloud/TalkAloud/Cells/TagCell/TagCell.swift [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    private var tagsDataSource = [TagModel]()
+    weak var delegate: TagFilterDelegate?
+    
+    override func awakeFromNib() {
+        super.awakeFromNib()
+       
+        self.collectionView.dataSource = self
+        self.collectionView.delegate = self
+        
+        self.collectionView.register(UINib(nibName: "TagCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: "tagCollectionViewCell")
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/Cells/TagCell/TagCell.swift]: Please change the reuseIdentifier to be "TagCollectionViewCell"

File: TalkAloud/TalkAloud/Cells/TagCell/TagCell.swift [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    
+    //==================================================
+    // MARK: - Collection View Data Source
+    //==================================================
+    
+    public func setup(with model: [TagModel]) {
+        self.tagsDataSource = model
+        self.collectionView.reloadData()
+    }
+
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/Cells/TagCell/TagCell.swift:40]: Remove this empty line

File: TalkAloud/TalkAloud/Cells/TagCell/TagCell.swift [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    
+    
+}
+
+//==================================================
+// MARK: - Collection View Data Source
+//==================================================
+
+extension TagCell: UICollectionViewDataSource {
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/Cells/TagCell/TagCell.swift:49]: And an empty line

File: TalkAloud/TalkAloud/Cells/TagCell/TagCell.swift [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        var selectedTag = tagsDataSource[indexPath.row]
+        selectedTag.isTagSelected = !selectedTag.isTagSelected
+        
+        tagsDataSource[indexPath.row] = selectedTag
+        
+        delegate?.didUpdateTagToFilter(by: tagsDataSource)
+        
+        self.collectionView.reloadData()
+    }
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/Cells/TagCell/TagCell.swift:75]: Add an empty line

File: TalkAloud/TalkAloud/Cells/TagCollectionViewCell/TagCollectionViewCell.swift [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    
+    public func setup(with model: TagModel) {
+        tagLabel.text = model.tag
+        
+        if model.isTagSelected {
+            self.backgroundColor = .red
+        } else {
+            self.backgroundColor = .black
+        }
+    }
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/Cells/TagCollectionViewCell/TagCollectionViewCell.swift:32]: Add an empty line

File: TalkAloud/TalkAloud/Cells/TagCollectionViewCell/TagCollectionViewCell.swift [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+//  TagCollectionViewCell.swift
+//  TalkAloud
+//
+//  Created by Justin Bengtson on 11/17/20.
+//  Copyright © 2020 Justin Bengtson. All rights reserved.
+//
+
+import UIKit
+
+class TagCollectionViewCell: UICollectionViewCell {
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/Cells/TagCollectionViewCell/TagCollectionViewCell.swift:11]: Please add the // MARK:

File: TalkAloud/TalkAloud/Managers/AudioManager.swift [hunk 9]
```diff
@@ -133,28 +134,58 @@ class AudioManager {
     //==================================================
     
     public func setTag(for selectedRecording: AudioRecording, tag: String) {
+
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/Managers/AudioManager.swift]: Remove this empty line

File: TalkAloud/TalkAloud/Managers/AudioManager.swift [hunk 10]
```diff
@@ -133,28 +134,58 @@ class AudioManager {
...
+                    }
                 }
             }
         }
-        allUniqueTags = allTags.unique
+    
+        return allTags
+    }
+    
+    public func addATag(tagModel: TagModel, tagModels: [TagModel]) -> [TagModel] {
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/Managers/AudioManager.swift]: Rename function to addTag
Also you shouldn't pass your dataSource from the ViewController to here. Modify this method so your AudioManager is the only source of truth. The data source from the AudioManager needs to "drive" the dataSource from the ViewController.

File: TalkAloud/TalkAloud/ViewControllers/AudioRecordingsViewController.swift [hunk 11]
```diff
@@ -29,20 +30,19 @@ class AudioRecordingsViewController: UIViewController {
...
     }
     
     //==================================================
-    // MARK: - Actions
+    // MARK: - Private Methods
     //==================================================
-    
-    @IBAction func tappedLeftButton(_ sender: Any) {
-        presentTagTableViewController()
+    private func getAllTags() -> [TagModel] {
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/ViewControllers/AudioRecordingsViewController.swift]: Why return smth from this function?

File: TalkAloud/TalkAloud/ViewControllers/AudioRecordingsViewController.swift [hunk 12]
```diff
@@ -71,8 +64,23 @@ class AudioRecordingsViewController: UIViewController {
...
+        var selectedTags = [String]()
+        
+        for tag in tags {
+            if tag.isTagSelected {
+                selectedTags.append(tag.tag)
+            }
+        }
+        
+        if selectedTags.isEmpty {
+            loadAudioRecordings(with: nil)
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/ViewControllers/AudioRecordingsViewController.swift]: Make this function to have the default value of the param be nil

File: TalkAloud/TalkAloud/ViewControllers/AudioRecordingsViewController.swift [hunk 13]
```diff
@@ -100,11 +108,28 @@ extension AudioRecordingsViewController: AudioRecordingCellDelegate {
 
 extension AudioRecordingsViewController: UITableViewDataSource {
     
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return 2
+    }
+    
     func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/ViewControllers/AudioRecordingsViewController.swift]: Remove empty line

File: TalkAloud/TalkAloud/ViewControllers/AudioRecordingsViewController.swift [hunk 14]
```diff
@@ -100,11 +108,28 @@ extension AudioRecordingsViewController: AudioRecordingCellDelegate {
...
+        
+        if section == 0 {
+            return 1
+        }
+        
         return audioRecordings.count
     }
     
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/ViewControllers/AudioRecordingsViewController.swift]: Remove empty line

File: TalkAloud/TalkAloud/Managers/AudioManager.swift [hunk 15]
```diff
@@ -133,28 +134,58 @@ class AudioManager {
...
+        var tagModelDataSource = tagModels
+        
+        if !tagModelDataSource.contains(tagModel) {
+            tagModelDataSource.append(tagModel)
+        }
+        
+        return tagModelDataSource
+    }
+    
+    public func removeTags(tags: [String], tagModels: [TagModel]) -> [TagModel] {
```
Comment (virgil-adumitroae) [inline TalkAloud/TalkAloud/Managers/AudioManager.swift]: Also you shouldn't pass your dataSource from the ViewController to here. Modify this method so your AudioManager is the only source of truth. The data source from the AudioManager needs to "drive" the dataSource from the ViewController.

**Review by virgil-adumitroae — CHANGES_REQUESTED** [review]

File: TalkAloud/TalkAloud/ViewControllers/AudioRecordingsViewController.swift (hunk 11 above)
Comment (JB184351) [reply TalkAloud/TalkAloud/ViewControllers/AudioRecordingsViewController.swift]: I was thinking that I needed to update the dataSource whenever I called this but after reading your comment I know I didn't need to do that.

**Pull request merged**